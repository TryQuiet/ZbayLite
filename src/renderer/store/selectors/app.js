import { createSelector } from 'reselect'

const store = s => s

export const app = createSelector(store, state => state.get('app'))

const version = createSelector(app, a => a.version)
const useTor = createSelector(app, a => a.useTor)
const transfers = createSelector(app, a => a.transfers)
const currentModalTab = createSelector(app, a => a.modalTabToOpen)
const allTransfersCount = createSelector(app, a => a.allTransfersCount)
const newTransfersCounter = createSelector(app, a => a.newTransfersCounter)
const isInitialLoadFinished = createSelector(app, a => a.isInitialLoadFinished)
const directMessageQueueLock = createSelector(
  app,
  a => a.directMessageQueueLock
)
const messageQueueLock = createSelector(app, a => a.messageQueueLock)

export default {
  version,
  transfers,
  currentModalTab,
  allTransfersCount,
  newTransfersCounter,
  messageQueueLock,
  directMessageQueueLock,
  isInitialLoadFinished,
  useTor
}
