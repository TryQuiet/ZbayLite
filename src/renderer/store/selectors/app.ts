import { createSelector } from 'reselect'

import { Store } from '../reducers'

const app = (s: Store) => s.app

const version = createSelector(app, (a) => a.version)
const currentModalTab = createSelector(app, (a) => a.modalTabToOpen)
const isInitialLoadFinished = createSelector(
  app,
  (a) => a.isInitialLoadFinished
)

export default {
  version,
  currentModalTab,
  isInitialLoadFinished
}
