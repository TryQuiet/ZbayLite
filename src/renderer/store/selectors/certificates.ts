import { Store } from '../reducers'
import { createSelector } from 'reselect'

const certificates = (s: Store) => s.certificates
const ownCertificate = createSelector(certificates, d => d.ownCertificate.certificate)

export default {
  certificates,
  ownCertificate
}
