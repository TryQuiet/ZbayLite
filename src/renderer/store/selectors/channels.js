import { createSelector } from 'reselect'

import zcashChannels from '../../zcash/channels'
import nodeSelectors from './node'

const ignoredChannels = network => [
  zcashChannels.registeredUsers[network].address,
  zcashChannels.channelOfChannels[network].address,
  zcashChannels.priceOracle[network].address
]
const store = s => s
const channels = createSelector(
  store,
  nodeSelectors.network,
  (state, network) => {
    return {
      ...state.channels,
      data: state.channels.data.filter(ch => !ignoredChannels(network).includes(ch.address))
    }
  }
)

const data = createSelector(channels, nodeSelectors.network, (ch, network) => {
  let channelsData = ch.data
  if (channelsData.length) {
    const storeId = channelsData.findIndex(
      ch => ch.address === zcashChannels.store[network].address
    )
    const storeChannel = channelsData.get(storeId)
    channelsData = channelsData.delete(storeId).unshift(storeChannel)
    const zbayId = channelsData.findIndex(
      ch => ch.address === zcashChannels.general[network].address
    )
    const zbayChannel = channelsData.get(zbayId)
    channelsData = channelsData.delete(zbayId).unshift(zbayChannel)
  }
  return channelsData
})
const loader = createSelector(channels, ch => ch.loader)
const errors = createSelector(channels, c => c.errors)

const generalChannelId = createSelector(
  data,
  nodeSelectors.network,
  (ch, network) => {
    return zcashChannels.general[network].address
  }
)
const usersChannel = createSelector(
  store,
  nodeSelectors.network,
  (state, network) => {
    const temp = {
      ...state.channels,
      data: state.channels.data.find(ch => ch.address === zcashChannels.registeredUsers[network].address)
    }
    return temp.data
  }
)
const priceOracleChannel = createSelector(
  store,
  nodeSelectors.network,
  (state, network) => {
    const temp = {
      ...state.channels,
      data: state.channels.data.find(ch => ch.address === zcashChannels.priceOracle[network].address)
    }
    return temp.data
  }
)
const publicChannels = createSelector(
  store,
  nodeSelectors.network,
  (state, network) => {
    const temp = {
      ...state.channels,
      data: state.channels.data.find(ch => ch.address === zcashChannels.channelOfChannels[network].address)
    }
    return temp.data
  }
)

const channelById = id =>
  createSelector(data, ch => ch.find(c => c.id === id))
const ownedChannels = createSelector(data, ch =>
  ch.filter(c => {
    return c.keys.sk
  })
)

const lastSeen = id => createSelector(channelById(id), ch => ch.lastSeen)

export default {
  channelById,
  generalChannelId,
  usersChannel,
  channels,
  loader,
  lastSeen,
  data,
  errors,
  ownedChannels,
  publicChannels,
  priceOracleChannel
}
