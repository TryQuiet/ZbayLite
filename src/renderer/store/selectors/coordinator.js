import { createSelector } from 'reselect'

const store = s => s

export const coordinator = createSelector(store, state => state.coordinator)

const running = createSelector(coordinator, a => a.running)

export default {
  running
}
