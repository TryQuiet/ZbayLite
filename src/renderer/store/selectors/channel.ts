import { createSelector } from 'reselect'
import identitySelectors from './identity'
import messagesQueueSelectors from './messagesQueue'
import contacts from './contacts'
import { networkFee, messageType } from '../../../shared/static'
import publicChannels from './publicChannels'
import users from './users'

import { Store } from '../reducers'
import { DisplayableMessage } from '../../zbay/messages.types'

const channel = (s: Store) => s.channel

export const channelInfo = createSelector(channel, ch => {
  const channel = {
    ...ch
  }
  delete channel.message
  return channel
}) // TODO refactor

const isPublicChannel = createSelector(
  publicChannels.publicChannels,
  channel,
  (pubChannels, channel) => {
    if (pubChannels && channel) {
      const publicChannelAddresses = Object.values(pubChannels).map(el => el.address)
      const { address } = channel
      return publicChannelAddresses.includes(address)
    } else {
      return false
    }
  }
)

export const spentFilterValue = createSelector(channel, c =>
  c.spentFilterValue ? spentFilterValue : -1
)

export const message = createSelector(channel, c => c.message[c.id] || '')
export const id = createSelector(channel, c => c.id)
const data = createSelector(contacts.contacts, id, (channels, id) => channels[id])
export const isSizeCheckingInProgress = createSelector(channel, c => c.isSizeCheckingInProgress)
export const messageSizeStatus = createSelector(channel, c => c.messageSizeStatus)
export const displayableMessageLimit = createSelector(channel, c => c.displayableMessageLimit)
export const isOwner = createSelector(
  id,
  contacts.contacts,
  identitySelectors.signerPubKey,
  (id, con, myKey) => {
    const contact = con[id]
    if (!contact) {
      return false
    }
    const settingsMsg = Array.from(Object.values(contact.messages)).filter(
      msg => msg.type === messageType.CHANNEL_SETTINGS
    )[0]
    if (settingsMsg && settingsMsg.message.owner === myKey) {
      return true
    }
    return false
  }
)
export const channelSettingsMessage = createSelector(data, data => {
  if (!data) {
    return null
  }
  const settingsMsg = Array.from(Object.values(data.messages)).filter(
    msg =>
      msg.type === messageType.CHANNEL_SETTINGS || msg.type === messageType.CHANNEL_SETTINGS_UPDATE
  )
  if (!settingsMsg.length) {
    return null
  }
  return settingsMsg.reduce((prev, curr) => (prev.createdAt > curr.createdAt ? prev : curr))
})
export const advertFee = createSelector(channelSettingsMessage, settingsMsg => {
  if (settingsMsg === null) {
    return 0
  }
  return settingsMsg.message.minFee || settingsMsg.message.updateMinFee
})
export const channelDesription = createSelector(channelSettingsMessage, settingsMsg => {
  if (settingsMsg === null) {
    return 0
  }
  return settingsMsg.message.updateChannelDescription || ''
})
export const onlyRegistered = createSelector(channelSettingsMessage, settingsMsg => {
  if (settingsMsg === null) {
    return 0
  }
  return settingsMsg.message.updateOnlyRegistered || '0'
})
export const unread = createSelector(data, data => (data ? data.unread : 0))

export const queuedMessages = createSelector(
  messagesQueueSelectors.queue,
  channel,
  (queue, channel) => queue.filter(m => m.channelId === channel.id)
)

export const loader = createSelector(channel, meta => meta.loader)

const checkMessageTargetTimeWindow = ({ targetCreatedAt, timeStamp, timeWindow }) => {
  const inRange = ({ timeStamp, targetCreatedAt, timeWindow }) => {
    return (timeStamp - targetCreatedAt) * (timeStamp - timeWindow) <= 0
  }
  return inRange({ timeStamp, targetCreatedAt, timeWindow })
}

const concatMessages = (mainMsg, messagesToConcat) => {
  if (messagesToConcat.length === 1) {
    return mainMsg
  } else {
    const messagesArray = messagesToConcat.map(msg => msg.message)
    const lastMessageStatus = messagesToConcat[messagesToConcat.length - 1].status
    const concatedMessages = messagesArray.join('\n')
    const mergedMessage = {
      ...mainMsg,
      message: concatedMessages,
      status: lastMessageStatus
    }
    return mergedMessage
  }
}

export const mergeIntoOne = (messages: DisplayableMessage[]) => {
  if (messages.length === 0) return
  const result = [[]]
  let last: DisplayableMessage = null
  for (const msg of messages) {
    const isMessageInTargetZone = last
      ? checkMessageTargetTimeWindow({
        targetCreatedAt: last.createdAt,
        timeStamp: msg.createdAt,
        timeWindow: last.createdAt + 300
      })
      : true
    if (last && msg.status === 'failed') {
      result.push([])
      result[result.length - 1].push(msg)
    } else if (last && (msg.type !== 1 || last.type !== 1)) {
      result.push([])
      result[result.length - 1].push(msg)
    } else if ((last && last.sender.replyTo !== msg.sender.replyTo) || !isMessageInTargetZone) {
      result.push([])
      result[result.length - 1].push(msg)
    } else {
      result[result.length - 1].push(msg)
    }
    last = msg
  }
  const concatedMessages = result.map(array => {
    return concatMessages(array[0], array)
  })
  return concatedMessages
}

export const shareableUri = createSelector(channel, c => c.shareableUri)

export const inputLocked = createSelector(
  identitySelectors.balance('zec'),
  identitySelectors.lockedBalance('zec'),
  users.users,
  identitySelectors.signerPubKey,
  (available, locked, users, signerPubKey) => {
    if (available.gt(networkFee)) {
      if (users[signerPubKey]) {
        if (users[signerPubKey].createdAt) {
          return INPUT_STATE.AVAILABLE
        } else {
          return INPUT_STATE.UNREGISTERED
        }
      }
    } else {
      if (locked.gt(0)) {
        return INPUT_STATE.LOCKED
      }
    }
    return INPUT_STATE.DISABLE
  }
)

export const INPUT_STATE = {
  DISABLE: 0,
  AVAILABLE: 1,
  LOCKED: 2,
  UNREGISTERED: 3
}

export const channelId = createSelector(channel, ch => ch.id)

export const members = createSelector(contacts.contacts, id, (c, channelId) => {
  const contact = c[channelId]
  if (!contact) {
    return new Set()
  }
  return Array.from(Object.values(contact.messages)).reduce((acc, msg) => {
    return acc.add(msg.sender.replyTo)
  }, new Set())
})

export const channelParticipiants = createSelector(contacts.contacts, id, (c, i) => {
  const contact = c[i]
  if (!contact) {
    return new Set()
  }
  const messages = contact.messages
  const members = Array.from(Object.values(messages)).reduce((acc, msg) => {
    return acc.add(msg.sender.replyTo)
  }, new Set())
  return members
})

export default {
  data,
  inputLocked,
  loader,
  channel,
  spentFilterValue,
  message,
  shareableUri,
  channelId,
  channelInfo,
  advertFee,
  onlyRegistered,
  members,
  channelParticipiants,
  unread,
  messageSizeStatus,
  isSizeCheckingInProgress,
  id,
  isOwner,
  channelDesription,
  displayableMessageLimit,
  isPublicChannel
}
