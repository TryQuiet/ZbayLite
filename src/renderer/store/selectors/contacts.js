import { createSelector } from 'reselect'
import Immutable from 'immutable'
import identitySelectors from './identity'
import directMssagesQueueSelectors from './directMessagesQueue'
import operationsSelectors from './operations'
import { operationTypes } from '../handlers/operations'
import usersSelectors from './users'
import { mergeIntoOne, displayableMessageLimit } from './channel'

export const Contact = Immutable.Record({
  lastSeen: null,
  key: '',
  username: '',
  address: '',
  messages: Immutable.Map(),
  newMessages: Immutable.List(),
  vaultMessages: Immutable.List(),
  offerId: null
})

const store = s => s

const contacts = createSelector(store, state => state.get('contacts'))
const contactsList = createSelector(
  contacts,
  identitySelectors.removedChannels,
  (contacts, removedChannels) => {
    if (removedChannels.size > 0) {
      return contacts
        .filter(
          c =>
            c.key.length === 66 &&
            c.offerId === null &&
            !removedChannels.includes(c.address)
        )
        .toList()
    }
    return contacts
      .filter(c => c.key.length === 66 && c.offerId === null)
      .toList()
  }
)

const offerList = createSelector(
  contacts,
  identitySelectors.removedChannels,
  (contacts, removedChannels) => {
    if (removedChannels.size > 0) {
      return contacts
        .filter(c => !!c.offerId && !removedChannels.includes(c.key))
        .toList()
    }
    return contacts.filter(c => !!c.offerId).toList()
  }
)
const channelsList = createSelector(
  contacts,
  identitySelectors.removedChannels,
  (contacts, removedChannels) => {
    if (removedChannels.size > 0) {
      return contacts
        .filter(
          c =>
            c.key.length === 78 &&
            c.offerId === null &&
            !removedChannels.includes(c.address)
        )
        .toList()
    }
    return contacts
      .filter(c => c.key.length === 78 && c.offerId === null)
      .toList()
  }
)

const directMessagesContact = address =>
  createSelector(contacts, c =>
    c.toList().find(el => el.get('address') === address)
  )

const contact = address =>
  createSelector(contacts, c => c.get(address, Contact()))
const messagesSorted = address =>
  createSelector(contact(address), c => {
    return c.messages
      .toList()
      .sort((msg1, msg2) => msg2.createdAt - msg1.createdAt)
  })
const messagesLength = address =>
  createSelector(contact(address), c => {
    return c.messages.toList().size
  })
const messages = address =>
  createSelector(
    messagesSorted(address),
    displayableMessageLimit,
    (msgs, limit) => {
      return msgs.slice(0, limit)
    }
  )

const allMessages = createSelector(contacts, c =>
  c.reduce((acc, t) => acc.merge(t.messages), Immutable.Map())
)
const getAdvertById = txid =>
  createSelector(allMessages, msgs => msgs.get(txid))
const lastSeen = address => createSelector(contact(address), c => c.lastSeen)
const username = address => createSelector(contact(address), c => c.username)
const vaultMessages = address =>
  createSelector(contact(address), c => c.vaultMessages)
const newMessages = address =>
  createSelector(contact(address), c => c.newMessages)

export const queuedMessages = address =>
  createSelector(
    directMssagesQueueSelectors.queue,
    queue =>
      queue.filter(
        m => m.recipientAddress === address && m.message.get('type') < 10
      ) //  separate offer messages and direct messages
  )

export const pendingMessages = address =>
  createSelector(operationsSelectors.operations, operations =>
    operations.filter(
      o =>
        o.type === operationTypes.pendingDirectMessage &&
        o.meta.recipientAddress === address &&
        o.meta.message.get('type') < 10 //  separate offer messages and direct messages
    )
  )

export const directMessages = (address, signerPubKey) =>
  createSelector(
    identitySelectors.data,
    usersSelectors.registeredUser(signerPubKey),
    messages(address),
    (identity, registeredUser, messages) => {
      // const userData = registeredUser ? registeredUser.toJS() : null
      // const identityAddress = identity.address
      // const identityName = userData ? userData.nickname : identity.name

      // const fetchedMessagesToDisplay = messages.map(msg =>
      //   zbayMessages.receivedToDisplayableMessage({
      //     message: msg,
      //     identityAddress,
      //     receiver: { replyTo: identityAddress, username: identityName }
      //   })
      // )

      const merged = mergeIntoOne(messages)
      return merged
    }
  )

export default {
  contacts,
  directMessagesContact,
  queuedMessages,
  pendingMessages,
  contact,
  messages,
  directMessages,
  lastSeen,
  vaultMessages,
  username,
  newMessages,
  contactsList,
  channelsList,
  offerList,
  getAdvertById,
  allMessages,
  messagesLength
}
