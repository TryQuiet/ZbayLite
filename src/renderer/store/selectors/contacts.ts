import { createSelector } from 'reselect'
import identitySelectors from './identity'
import usersSelectors from './users'
import publicChannelsSelectors from './publicChannels'
import directMssagesQueueSelectors from './directMessagesQueue'
import { mergeIntoOne, displayableMessageLimit } from './channel'
import { MessageType } from '../../../shared/static.types'
import { unknownUserId } from '../../../shared/static'

import { DisplayableMessage } from '../../zbay/messages.types'

import { Contact } from '../handlers/contacts'
import { Store } from '../reducers'
import certificatesSelector from '../certificates/certificates.selector'
import { extractPubKeyString } from '../../pkijs/tests/extractPubKey'
import { loadCertificate } from '../../pkijs/generatePems/common'
import channelSelector from '../selectors/channel'

const contacts = (s: Store) => s.contacts

const contactExists = (address: string) => createSelector(contacts, allContacts => {
  return Object.keys(allContacts).includes(address)
})

const publicChannelsContacts = createSelector(contacts, publicChannelsSelectors.publicChannels, (allContacts, publicChannels) => {
  const pChannels = Object.values(publicChannels).map(pc => pc.address)
  return Object.values(allContacts).filter(contact => pChannels.includes(contact.address))
})

const contactsList = createSelector(
  contacts,
  identitySelectors.removedChannels,
  usersSelectors.users,
  (contacts, removedChannels, users) => {
    return Array.from(Object.values(contacts))
      .map(contact => {
        const user = users[contact.key]
        if (!contact.address && user) {
          return {
            ...contact,
            messages: messages,
            address: user.address || contact.address,
            username: user.nickname || contact.username
          }
        } else {
          return contact
        }
      })
      .filter(
        c => c.key.length === 66 && c.offerId === null && !removedChannels.includes(c.address)
      )
  }
)

const unknownMessages = createSelector(contacts, contacts => {
  return Array.from(Object.values(contacts)).filter(c => c.key === unknownUserId)
})

const offerList = createSelector(
  contacts,
  identitySelectors.removedChannels,
  (contacts, removedChannels) => {
    if (removedChannels.length > 0) {
      return Array.from(Object.values(contacts)).filter(
        c => !!c.offerId && !removedChannels.includes(c.key)
      )
    }
    return Array.from(Object.values(contacts)).filter(c => !!c.offerId)
  }
)
const channelsList = createSelector(
  contacts,
  identitySelectors.removedChannels,
  (contacts, removedChannels) => {
    return Array.from(Object.values(contacts)).filter(
      c => c.key.length === 78 && c.offerId === null && !removedChannels.includes(c.address)
    )
  }
)

const directMessagesContact = address =>
  createSelector(contacts, c => Array.from(Object.values(c)).find(el => el.address === address))

const contact = address =>
  createSelector(contacts, usersSelectors.users, (c, u) => {
    if (!c[address]) {
      return new Contact()
    } else {
      if (!c[address].address) {
        return {
          ...c[address],
          address: u[address]?.address,
          username: u[address]?.nickname || c[address].username
        }
      } else {
        return c[address]
      }
    }
  })

const messagesSorted = address =>
  createSelector(contact(address), usersSelectors.users, (c, u) => {
    return Array.from(Object.values(c.messages))
      .sort((a, b) => b.createdAt - a.createdAt)
      .map(message => {
        if (message.isUnregistered) {
          return {
            ...message,
            sender: {
              ...message.sender,
              username: u[message.pubKey]?.nickname || message.sender.username
            }
          }
        } else {
          return message
        }
      })
  })
const messagesSortedDesc = address =>
  createSelector(contact(address), c => {
    return Array.from(Object.values(c.messages)).sort((a, b) => a.createdAt - b.createdAt)
  })

const messagesLength = address =>
  createSelector(contact(address), c => {
    return Array.from(Object.values(c.messages)).length
  })
const messages = address =>
  createSelector(messagesSorted(address), displayableMessageLimit, (msgs, limit) => {
    return msgs.slice(0, limit)
  })

const channelSettingsMessages = address =>
  createSelector(messagesSortedDesc(address), msgs => {
    return msgs.filter(msg => msg.type === 6)
  })
const channelModerators = address =>
  createSelector(directMessages(address), msgs => {
    return msgs.channelModerators
  })

const allMessages = createSelector(contacts, c => {
  return Array.from(Object.keys(c)).reduce((acc, t) => {
    const temp = (acc[t] = {
      ...acc,
      ...c[t].messages
    })
    return temp
  }, {})
})
const allMessagesTxnId = createSelector(allMessages, c => {
  return new Set(Object.keys(c))
})
const getAdvertById = (txid: string) =>
  createSelector(allMessages, msgs => {
    return msgs[txid]
  })
const lastSeen = address => createSelector(contact(address), c => c.lastSeen)
const username = address => createSelector(contact(address), c => c.username)
const vaultMessages = address => createSelector(contact(address), c => c.vaultMessages)
const newMessages = address => createSelector(contact(address), c => c.newMessages)

export const queuedMessages = address =>
  createSelector(
    directMssagesQueueSelectors.queue,
    queue => queue.filter(m => m.recipientAddress === address && m.message.type < 10) //  separate offer messages and direct messages
  )

const channelOwner = channelId =>
  createSelector(channelSettingsMessages(channelId), msgs => {
    let channelOwner = null
    channelOwner = msgs[0] ? msgs[0].pubKey : null
    for (const msg of msgs) {
      if (channelOwner === msg.pubKey) {
        channelOwner = msg.owner
      }
    }
    return channelOwner
  })
// TODO: TO be removed
export interface IDirectMessage {
  visibleMessages: DisplayableMessage[]
  channelModerators: string[]
  messsagesToRemove: DisplayableMessage[]
  blockedUsers: string[]
}

const currentChannel = createSelector(
  contacts,
  channelSelector.address,
  channelSelector.id,
  (contacts, address, id) => {
    if (contacts[address]) {
      return contacts[address]
    } else {
      return contacts[id]
    }
  }
)

const usersCertificateMapping = createSelector(
  certificatesSelector.usersCertificates,
  (certificates) => {
    return certificates.reduce<{ [pubKey: string]: { username: string; onionAddress: string; peerId: string } }>((acc, current) => {
      let parsedCerficated
      let certObject
      let nickname = null
      let onionAddress = null
      let peerId = null
      if (current !== null && current) {
        parsedCerficated = extractPubKeyString(current)
        certObject = loadCertificate(current)
        if (certObject.subject.typesAndValues.length === 3) {
          nickname = certObject.subject.typesAndValues[0].value.valueBlock.value
          onionAddress = certObject.subject.typesAndValues[1].value.valueBlock.value
          peerId = certObject.subject.typesAndValues[2].value.valueBlock.value
        } else {
          return
        }
      }
      acc[parsedCerficated] = {
        username: nickname,
        onionAddress: onionAddress,
        peerId: peerId
      }
      return acc
    }, {})
  }
)

const messagesOfChannelWithUserInfo = createSelector(
  currentChannel, usersCertificateMapping,
  (currentChannel, usersCertificateMapping) => {
    if (!currentChannel) return []
    const messagesArray = Object.values(currentChannel.messages)
    return messagesArray.map(
      message => {
        if (usersCertificateMapping[message.pubKey] && message.id !== '44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a') {
          const userInfo = usersCertificateMapping[message.pubKey]
          if (userInfo.onionAddress !== null) {
            return ({ message, userInfo: userInfo })
          }
        }
      }
    ).filter((item) => item !== undefined)
  }
)

export const directMessages = address => createSelector(
  messagesOfChannelWithUserInfo,
  channelOwner(address),
  (messagesWithUserInfo, channelOwner): IDirectMessage => {
    const messagesObjectsArray = messagesWithUserInfo.map((message) => {
      const newMessage = {
        ...message.message,
        createdAt: Math.floor(message.message.createdAt),
        sender: {
          username: message.userInfo ? message.userInfo.username : 'unNamed',
          replyTo: ''
        }
      }
      return newMessage
    })
    const sortedMessages = messagesObjectsArray.sort((a, b) => {
      return b.createdAt - a.createdAt
    })
      .map(message => message)

    const messages = sortedMessages

    const channelModerators = []
    const messsagesToRemove: DisplayableMessage[] = []
    const blockedUsers = []
    let visibleMessages: DisplayableMessage[] = []
    for (const msg of messages.reverse()) {
      switch (msg.type) {
        case MessageType.AD:
          if (!blockedUsers.includes(msg.pubKey)) {
            visibleMessages.push(msg)
          }
          break
        case MessageType.BASIC:
          if (!blockedUsers.includes(msg.pubKey)) {
            visibleMessages.push(msg)
          }
          break
        case MessageType.TRANSFER:
          if (!blockedUsers.includes(msg.pubKey)) {
            visibleMessages.push(msg)
          }
          break
        case MessageType.MODERATION:
          const senderPk = msg.pubKey
          const moderationType = msg.message.moderationType
          const moderationTarget = msg.message.moderationTarget
          if (channelOwner === senderPk && moderationType === 'ADD_MOD') {
            channelModerators.push(moderationTarget)
          } else if (channelOwner === senderPk && moderationType === 'REMOVE_MOD') {
            const indexToRemove = channelModerators.findIndex(el => el === moderationTarget)
            if (indexToRemove !== -1) {
              channelModerators.splice(indexToRemove, 1)
            }
          } else if (
            (channelOwner === senderPk || channelModerators.includes(senderPk)) &&
            moderationType === 'BLOCK_USER'
          ) {
            blockedUsers.push(moderationTarget)
            visibleMessages = visibleMessages.filter(msg => !blockedUsers.includes(msg.pubKey))
          } else if (
            (channelOwner === senderPk || channelModerators.includes(senderPk)) &&
            moderationType === 'UNBLOCK_USER'
          ) {
            const indexToRemove = blockedUsers.findIndex(el => el === moderationTarget)
            if (indexToRemove !== -1) {
              blockedUsers.splice(indexToRemove, 1)
            }
          } else if (
            (channelOwner === senderPk || channelModerators.includes(senderPk)) &&
            moderationType === 'REMOVE_MESSAGE'
          ) {
            const indexToRemove = visibleMessages.findIndex(el => el.id === moderationTarget)
            if (indexToRemove !== -1) {
              visibleMessages.splice(indexToRemove, 1)
            }
          } else {
          }
          break
      }
    }
    const result: IDirectMessage = {
      channelModerators,
      messsagesToRemove,
      blockedUsers,
      visibleMessages: mergeIntoOne(visibleMessages.reverse())
    }
    return result
  }
)

export default {
  contacts,
  contactExists,
  publicChannelsContacts,
  directMessagesContact,
  queuedMessages,
  channelModerators,
  contact,
  messages,
  directMessages,
  lastSeen,
  vaultMessages,
  username,
  newMessages,
  contactsList,
  channelsList,
  offerList,
  getAdvertById,
  allMessages,
  messagesLength,
  messagesSorted,
  unknownMessages,
  allMessagesTxnId
}
