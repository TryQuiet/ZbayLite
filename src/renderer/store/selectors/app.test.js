/* eslint import/first: 0 */
import selectors from './app'
import { initialState as AppState } from '../handlers/app'

import create from '../create'

describe('app -', () => {
  let store = null
  beforeEach(() => {
    jest.clearAllMocks()
    store = create({
      app: {
        ...AppState,
        version: '0.13.37',
        transfers: {},
        modalTabToOpen: 'addFunds',
        allTransfersCount: 12,
        newTransfersCounter: 2
      }
    })
  })

  it('version selector', async () => {
    expect(selectors.version(store.getState())).toMatchSnapshot()
  })
  it('version transfers', async () => {
    expect(selectors.transfers(store.getState())).toMatchSnapshot()
  })
  it('modal tab to open', async () => {
    expect(selectors.currentModalTab(store.getState())).toMatchSnapshot()
  })
  it('allTransfersCount', async () => {
    expect(selectors.allTransfersCount(store.getState())).toMatchSnapshot()
  })
  it('newTransfersCounter', async () => {
    expect(selectors.newTransfersCounter(store.getState())).toMatchSnapshot()
  })
  it('messageQueueLock', async () => {
    expect(selectors.messageQueueLock(store.getState())).toMatchSnapshot()
  })
  it('directMessageQueueLock', async () => {
    expect(selectors.directMessageQueueLock(store.getState())).toMatchSnapshot()
  })
  it('isInitialLoadFinished', async () => {
    expect(selectors.isInitialLoadFinished(store.getState())).toMatchSnapshot()
  })
})
