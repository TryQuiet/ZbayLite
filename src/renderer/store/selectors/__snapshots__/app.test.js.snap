// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app - allTransfersCount 1`] = `12`;

exports[`app - directMessageQueueLock 1`] = `false`;

exports[`app - isInitialLoadFinished 1`] = `false`;

exports[`app - messageQueueLock 1`] = `false`;

exports[`app - modal tab to open 1`] = `"addFunds"`;

exports[`app - newTransfersCounter 1`] = `2`;

exports[`app - version selector 1`] = `"0.13.37"`;

exports[`app - version transfers 1`] = `Object {}`;
